gbus/bus.go

```diff
@@ -488,23 +489,67 @@ func (b *DefaultBus) sendWithTx(ctx context.Context, ambientTx *sql.Tx, toServic
 
 func (b *DefaultBus) returnDeadToQueue(ctx context.Context, ambientTx *sql.Tx, publishing *amqp.Publishing) error {
 	if !b.started {
-		return errors.New("bus not strated or already shutdown, make sure you call bus.Start() before sending messages")
+		return errors.New("bus not started or already shutdown, make sure you call bus.Start() before sending messages")
+	}
+
+	targetQueue, ok := publishing.Headers["x-first-death-queue"].(string)
+	if !ok {
+		return fmt.Errorf("bad x-first-death-queue field - %v", publishing.Headers["x-first-death-queue"])
+	}
+	exchange, ok := publishing.Headers["x-first-death-exchange"].(string)
+	if !ok {
+		return fmt.Errorf("bad x-first-death-exchange field - %v", publishing.Headers["x-first-death-exchange"])
+	}
+	routingKey, err := extractRoutingKey(publishing.Headers)
+	if err != nil {
+		return err
 	}
-	//publishing.Headers.
-	exchange := fmt.Sprintf("%v", publishing.Headers["x-first-death-exchange"])
-	routingKey := fmt.Sprintf("%v", publishing.Headers["x-first-death-queue"])
+
+	publishing.Headers[ResurrectedHeaderName] = true
+	publishing.Headers["x-first-death-routing-key"] = routingKey
+	// publishing.Headers["x-first-death-exchange"] is not deleted and kept as is
 
 	delete(publishing.Headers, "x-death")
 	delete(publishing.Headers, "x-first-death-queue")
 	delete(publishing.Headers, "x-first-death-reason")
-	delete(publishing.Headers, "x-first-death-exchange")
+
+	b.Log().
```

Comment:
I would add the message_id to the fields as well so we can have visibility across all of the lifecycle of the message

grabbit -> handle -> reject -> deadletter -> replay -> grabbit -> handle